gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js

```diff
@@ -0,0 +1,129 @@
+    prefsPromise.then(request.events.onSuccess.fire, request.events.onError.fire);
+};
+
+fluid.defaults("gpii.preferencesServer.prefsSafeList.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.preferencesServer.prefsSafeList.handler.getPrefsSafeList",
+            args: ["{gpii.preferencesServer.preferencesService}", "{preferencesServer}", "{that}", "{that}.req.query.view"]
+        }
+    }
+});
+
+gpii.preferencesServer.prefsSafeList.handler.getPrefsSafeList = function (prefsService, preferencesServer, request, view) {
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listPrefsSafesDataSource.options.directModel, {});
+    var prefsPromise = prefsService.dataStore.listPrefsSafesDataSource.get(finalDirectModel);
+    prefsPromise.then(function (data) {
+        var togo = [];
+        // TODO this should probably be done in a couchview for efficiency and eventual paging?
+        fluid.each(data, function(item) {
+            item.value.prefsSafeId = item.key;
+            togo.push(item.value);
+        });
+        request.events.onSuccess.fire(togo);
+    }, request.events.onError.fire);
+};
+
+fluid.defaults("gpii.preferencesServer.listKeysForPrefsSafe.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.preferencesServer.listKeysForPrefsSafe.handler.getKeysForPrefsSafe",
+            args: ["{gpii.preferencesServer.preferencesService}", "{preferencesServer}", "{that}", "{that}.req.params.prefsSafeId", "{that}.req.query.view"]
+        }
+    }
+});
+
+gpii.preferencesServer.listKeysForPrefsSafe.handler.getKeysForPrefsSafe = function (prefsService, preferencesServer, request, prefsSafeId, view) {
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel, {
+        prefsSafeId: prefsSafeId});
+    var prefsPromise = prefsService.dataStore.listKeysForPrefsSafeDataSource.get(finalDirectModel);
+    prefsPromise.then(function (data) {
+        var togo = [];
+        fluid.each(data, function (i) { 
+            var next = i.doc;
+            gpii.dbOperation.dbDataStore.cleanUpDoc(next);
+            togo.push(next);
+        });
+        request.events.onSuccess.fire(togo);
+    }, request.events.onError.fire);
+};
+
+fluid.defaults("gpii.preferencesServer.prefsSafeWithKeysGet.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.preferencesServer.prefsSafeWithKeysGet.handler.get",
+            args: ["{gpii.preferencesServer.preferencesService}", "{preferencesServer}", "{that}", "{that}.req.params.prefsSafeId", "{that}.req.query.view"]
+        }
+    }
+});
+
+gpii.preferencesServer.prefsSafeWithKeysGet.handler.get = function (prefsService, preferencesServer, request, prefsSafeId, view) {
```

Comment:
This function does 2 things:
1. Find keys associated with a prefs safe;
2. Find the data of this prefs safe and add keys retrieved at step 1 to the data.

A couple of suggestions:
1. The step 1 is shared by this handler and `gpii.preferencesServer.listKeysForPrefsSafe.handler`, so it can be extracted into a function to get rid of the duplication;
2. Use [`fluid.promise.fireTransformEvent()`](https://docs.fluidproject.org/infusion/development/PromisesAPI.html#fluidpromisefiretransformeventevent-payload-options) to chain these 2 promises that the output of the first promise can be passed to the second promise. Using `fluid.promise.fireTransformEvent()` also helps to reuse the function in the first suggestion.